File: source/jugglinglab/notation/Pattern.java
// Pattern.java
//
// Copyright 2002-2022 Jack Boyce and the Juggling Lab contributors

package jugglinglab.notation;

import java.util.ResourceBundle;

import jugglinglab.jml.JMLPattern;
import jugglinglab.util.*;


// This is the base class for all non-JML pattern types in Juggling Lab.
// It parses from a string representation and creates a JMLPattern version of
// itself for the animator.

public abstract class Pattern {
    static final ResourceBundle guistrings = jugglinglab.JugglingLab.guistrings;
    static final ResourceBundle errorstrings = jugglinglab.JugglingLab.errorstrings;

    // The built-in notations
    public static final String[] builtinNotations = { "Siteswap" };

    // these should be in the same order as in the builtinNotations array
    public static final int NOTATION_NONE = 0;
    public static final int NOTATION_SITESWAP = 1;

    // creates a new blank pattern in the given notation
    public static Pattern newPattern(String notation) throws JuggleExceptionUser,
                                                JuggleExceptionInternal {
        if (notation == null)
            throw new JuggleExceptionUser("Notation type not specified");

        if (notation.equalsIgnoreCase("siteswap"))
            return new SiteswapPattern();

        throw new JuggleExceptionUser("Notation type '"+notation+"' not recognized");
    }

    // return the notation name
    public abstract String getNotationName();

    // define pattern from textual representation
    public abstract Pattern fromString(String config) throws
                        JuggleExceptionUser, JuggleExceptionInternal;

    // define pattern from ParameterList input
    public abstract Pattern fromParameters(ParameterList pl) throws
                        JuggleExceptionUser, JuggleExceptionInternal;

    // canonical string representation
    public abstract String toString();

    // convert pattern to JML
    public abstract JMLPattern asJMLPattern() throws
                        JuggleExceptionUser, JuggleExceptionInternal;
}

File: composeApp/src/commonMain/kotlin/jugglinglab/notation/Pattern.kt
//
// Pattern.kt
//
// This is the base class for all non-JML pattern types in Juggling Lab.
// It parses from a string representation and creates a JmlPattern version of
// itself for the animator.
//
// Copyright 2002-2026 Jack Boyce and the Juggling Lab contributors
//

package jugglinglab.notation

import jugglinglab.jml.JmlPattern
import jugglinglab.util.JuggleExceptionInternal
import jugglinglab.util.JuggleExceptionUser
import jugglinglab.util.ParameterList

abstract class Pattern {
    // return the notation name
    abstract val notationName: String

    // define pattern from textual representation
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun fromString(config: String): Pattern

    // define pattern from ParameterList input
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun fromParameters(pl: ParameterList): Pattern

    // canonical string representation
    abstract override fun toString(): String

    // convert pattern to JML
    @Throws(JuggleExceptionUser::class, JuggleExceptionInternal::class)
    abstract fun asJmlPattern(): JmlPattern

    @Suppress("unused")
    companion object {
        // The built-in notations
        val builtinNotations: List<String> = listOf(
            "Siteswap",
            )

        // these should be in the same order as in the builtinNotations array
        const val NOTATION_NONE: Int = 0
        const val NOTATION_SITESWAP: Int = 1

        // Create a new blank pattern in the given notation.

        @Throws(JuggleExceptionUser::class)
        fun newPattern(notation: String): Pattern {
            if (notation.equals("siteswap", ignoreCase = true)) {
                return SiteswapPattern()
            }
            throw JuggleExceptionUser("Notation type '$notation' not recognized")
        }

        // Return the notation name with canonical capitalization.

        fun canonicalNotation(notation: String?): String? {
            for (n in builtinNotations) {
                if (n.equals(notation, ignoreCase = true)) {
                    return n
                }
            }
            return null
        }
    }
}
